starfish/util/overlap_utils.py

```diff
@@ -0,0 +1,169 @@
+import itertools
+from typing import List, Optional, Tuple
+
+import xarray as xr
+
+from starfish.types import Coordinates, Features, Number
+from starfish.types._constants import OverlapStrategy
+
+
+class Area:
+    """
+    Small class that defines rectangular area of physical space by
+    its bottom left and top right coordinates.
+    """
+    def __init__(self, min_x: Number, max_x: Number, min_y: Number, max_y: Number):
+        self.min_x = min_x
+        self.max_x = max_x
+        self.min_y = min_y
+        self.max_y = max_y
+
+    def __eq__(self, other):
+        return (self.min_x == other.min_x
+                and self.min_y == other.min_y
+                and self.max_x == other.max_x
+                and self.max_y == other.max_y)
+
+    @staticmethod
+    def _overlap(area1: "Area", area2: "Area") -> bool:
+        """Return True if two rectangles overlap"""
+        if (area1.max_x < area2.min_x) or (area1.min_x > area2.max_x):
+            return False
+        if (area1.max_y < area2.min_y) or (area1.min_y > area2.max_y):
+            return False
+        return True
+
+    @staticmethod
+    def find_intersection(area1: "Area", area2: "Area") -> Optional["Area"]:
+        """
+        Find the overlap area of two rectangles and return as new Area object.
+        If the two rectangles do not overlap, return None.
+        """
+        if Area._overlap(area1, area2):
+            return Area(max(area1.min_x, area2.min_x),
+                        min(area1.max_x, area2.max_x),
+                        max(area1.min_y, area2.min_y),
+                        min(area1.max_y, area2.max_y))
+        return None
+
+
+def find_overlaps_of_xarrays(xarrays: List[xr.DataArray]) -> List[Tuple[int, int]]:
+    """
+    Find all the overlap areas within a list of xarrays.
+
+    Parameters
+    ----------
+    xarrays : List[xr.DataArray]
+        The list of xarrays to find overlaps in.
+
+    Returns
+    -------
+    list[Tuple[int, int]] :
+        A list of tuples containing the indices of two overlapping
+        IntensityTables.
+
+    """
+    all_overlaps: List[Tuple[int, int]] = list()
+    for idx1, idx2 in itertools.combinations(range(len(xarrays)), 2):
```

Comment:
so this is, as you pointed out, a n^2 operation.  but each operation requires scanning the table to find the min/max for the coordinates.  can you get some numbers from @ambrosejcarr for realistic FOV and spot counts, and build a set of tables that reflect that, and see what the perf is like?  we can likely significantly shrink the cost by precomputing a min-max for each xarray and reusing that.  it would also help inform whether we need to do the nlogn approach.

Finally, it might be worth trying to actually merge a large set of intensity tables, even if synthetic, to see if there are any performance implications to any of the xarray ops used during the merge.